geth/node/manager.go

```diff
@@ -469,31 +486,270 @@ func (m *NodeManager) AccountKeyStore() (*keystore.KeyStore, error) {
+	m.nodeStartedLock.RLock()
+	defer m.nodeStartedLock.RUnlock()
+
+	return m.nodeStarted
+}
+
+func (m *NodeManager) closeNodeStarted() {
+	m.nodeStartedLock.Lock()
+	close(m.nodeStarted)
+	m.nodeStartedLock.Unlock()
+}
+
+func (m *NodeManager) waitNodeStarted() {
+	m.nodeStartedLock.RLock()
+	<-m.nodeStarted
+	m.nodeStartedLock.RUnlock()
+}
+
+// setStarted sets node into started state: start channel closed, and send Started signal.
+func (m *NodeManager) setStarted() {
+	if !m.isNodePending() {
+		return
+	}
+
+	m.closeNodeStarted()
+	m.setStartedState()
+
+	signal.Send(signal.Envelope{
+		Type:  signal.EventNodeStarted,
+		Event: struct{}{},
+	})
+}
+
+// setStopped sets node into stopped state: start and stop channels closed, and send Stopped signal.
+func (m *NodeManager) setStopped() {
+	if m.isNodeStopped() {
+		return
+	}
+
+	log.Info("Node is stopped")
+
+	m.node.Wait()
+	m.closeNodeStopped()
+	m.setStoppedState()
+
+	// notify application that it can send more requests now
+	signal.Send(signal.Envelope{
+		Type:  signal.EventNodeStopped,
+		Event: struct{}{},
+	})
+
+	log.Info("Node manager notified app, that node has stopped")
+}
+
+// setFailed sets node into Failed state: stopped state, start and stop channels closed, and send Crash signal.
+func (m *NodeManager) setFailed(err error) {
+	if !m.isNodePending() {
+		return
+	}
+
+	m.nodeStartedLock.Lock()
+	close(m.nodeStarted)
+	m.nodeStartedLock.Unlock()
+
+	m.setStoppedState()
+
+	signal.Send(signal.Envelope{
+		Type: signal.EventNodeCrashed,
+		Event: signal.NodeCrashEvent{
+			Error: err.Error(),
+		},
+	})
+}
+
+func (m *NodeManager) getNodeStopped() chan struct{} {
+	m.nodeStoppedLock.RLock()
+	defer m.nodeStoppedLock.RUnlock()
+
+	return m.nodeStopped
+}
+
+func (m *NodeManager) setNodeStopped(nodeStopped chan struct{}) {
+	m.nodeStoppedLock.Lock()
+	m.nodeStopped = nodeStopped
+	m.nodeStoppedLock.Unlock()
+}
+
+func (m *NodeManager) closeNodeStopped() {
+	m.nodeStoppedLock.Lock()
+	close(m.nodeStopped)
+	m.nodeStoppedLock.Unlock()
+}
+
+// getWhisperService returns Whisper service or inits it
+func (m *NodeManager) getWhisperService() (*whisperv5.Whisper, error) {
+	m.whisperServiceLock.Lock()
+	defer m.whisperServiceLock.Unlock()
+
+	if m.whisperService != nil {
+		return m.whisperService, nil
+	}
+
+	var whisperObject *whisperv5.Whisper
+	err := m.node.Service(&whisperObject)
+	if err != nil {
+		return nil, ErrInvalidWhisperService
+	}
+	m.whisperService = whisperObject
+
+	return m.whisperService, nil
+}
+
+// getLesService returns LES service or inits both LES and backend
+func (m *NodeManager) getLesService() (services.LesService, error) {
+	m.lesServiceLock.Lock()
+	defer m.lesServiceLock.Unlock()
+
+	if m.lesService != nil {
+		return m.lesService, nil
+	}
+
+	var lesObject *les.LightEthereum
+	err := m.node.Service(&lesObject)
+	if err != nil {
+		return nil, ErrInvalidLightEthereumService
+	}
+	m.lesService = lesObject
+
+	m.statusBackend = lesObject.StatusBackend
```

Comment:
Definitely not a good idea that a method called `getLesService` assigns `m.statusBackend = lesObject.StatusBackend`. It will result in a reference cycle `NodeManager -> StatusBackend -> NodeManager`.

Reply:
I`m not sure what do you mean `reference cyrcle`:
StatusBackend doesnt contain NodeManager or reference to it.

Reply:
Oh because it's a different `StatusBackend` from the geth. As you can see it's confusing :) I would try not to assign `StatusBackend` from `les` anywhere. It's gonna be removed eventually.